python-checks/src/main/java/org/sonar/python/checks/SpecialMethodReturnTypeCheck.java

```diff
@@ -0,0 +1,361 @@
+      return InferredTypes.NONE;
+    }
+
+    if (returnedExpressions.size() == 1) {
+      return returnedExpressions.get(0).type();
+    }
+
+    return InferredTypes.TUPLE;
+  }
+
+  /**
+   * Calls {@code ctx.addIssue} for a return statement such that...
+   *
+   * ...all returned expressions are marked as the source of the issue if the return statement contains such expressions
+   * ...the return keyword is marked as the source of the issue if the return statement does not contain any expressions
+   */
+  private static void addIssueOnReturnedExpressions(SubscriptionContext ctx, ReturnStatement returnStatement, String message) {
+    final List<Expression> returnedExpressions = returnStatement.expressions();
+
+    if (returnedExpressions.isEmpty()) {
+      ctx.addIssue(returnStatement.returnKeyword(), message);
+    } else {
+      final Token firstExpressionToken = returnedExpressions.get(0).firstToken();
+      final Token lastExpressionToken = returnedExpressions.get(returnedExpressions.size() - 1).lastToken();
+
+      ctx.addIssue(firstExpressionToken, lastExpressionToken, message);
+    }
+  }
+
+  private static ReturnStmtCollector collectReturnStmts(FunctionDef funDef) {
+    final ReturnStmtCollector returnExpressionCollector = new ReturnStmtCollector();
+    funDef.body().accept(returnExpressionCollector);
+
+    return returnExpressionCollector;
+  }
+
+  private static class ReturnStmtCollector extends BaseTreeVisitor {
+    private final List<ReturnStatement> returnStmts = new ArrayList<>();
+    private List<Token> yieldKeywords = new ArrayList<>();
+    private boolean raisesWhitelistingException = false;
+
+    public List<ReturnStatement> getReturnStmts() {
+      return returnStmts;
+    }
+
+    public List<Token> getYieldKeywords() {
+      return yieldKeywords;
+    }
+
+    public boolean raisesWhitelistingException() {
+      return raisesWhitelistingException;
+    }
+
+    @Override
+    public void visitReturnStatement(ReturnStatement pyReturnStatementTree) {
+      returnStmts.add(pyReturnStatementTree);
+    }
+
+    @Override
+    public void visitFunctionDef(FunctionDef pyFunctionDefTree) {
+      // We do not visit nested function definitions as they may contain irrelevant return statements
+    }
+
+    @Override
+    public void visitYieldStatement(YieldStatement pyYieldStatementTree) {
+      yieldKeywords.add(pyYieldStatementTree.yieldExpression().yieldKeyword());
+    }
+
+    @Override
+    public void visitYieldExpression(YieldExpression pyYieldExpressionTree) {
+      yieldKeywords.add(pyYieldExpressionTree.yieldKeyword());
+    }
+
+    @Override
+    public void visitLambda(LambdaExpression pyLambdaExpressionTree) {
+      // We do not visit nested lambda definitions as they may contain irrelevant yield expressions
+    }
+
+    @Override
+    public void visitRaiseStatement(RaiseStatement pyRaiseStatementTree) {
+      raisesWhitelistingException |= pyRaiseStatementTree
```

Comment:
Looking at ruling issues, I think it probably makes sense to avoid reporting an issue when the method implementation has only "raise statements" as exit points, regardless of the exception raised. This should also simplify the logic here below.

E.g.

```python
    def __hash__(self):
        self.hashCalled = True
        raise RuntimeError('not a TypeError!') # Compliant - done intentionally
```